app/PaymentDrivers/Blockonomics/Blockonomics.php

```diff
@@ -132,20 +140,52 @@ public function paymentResponse(PaymentResponseRequest $request)
             'currency' => ['required'],
             'txid' => ['required'],
             'payment_method_id' => ['required'],
+            'btc_address' => ['required'],
+            'btc_amount' => ['required'],
+            'btc_price' => ['required'],
+            // Setting status to required will break the payment process
+            // because sometimes the status is returned as 0 which is falsy
+            // and the validation will fail.
+            // 'status' => ['required'],
         ]);
 
         try {
             $data = [];
-            $fiat_amount = round(($request->btc_price * $request->btc_amount), 2);
+            $fiat_amount = round(($request->btc_price * $request->btc_amount / 100000000), 2);
             $data['amount'] = $fiat_amount;
-            $data['currency'] = $request->currency;
             $data['payment_method_id'] = $request->payment_method_id;
             $data['payment_type'] = PaymentType::CRYPTO;
             $data['gateway_type_id'] = GatewayType::CRYPTO;
-            $data['transaction_reference'] = $request->txid;
 
-            $statusId = Payment::STATUS_PENDING;
+            // Append a random value to the transaction reference for test payments
+            // to prevent duplicate entries in the database.
+            // This ensures the payment hashed_id remains unique.
+            $testTxid = $this->getTestTxid();
+            $data['transaction_reference'] = ($request->txid === $testTxid)
+                ? $request->txid . bin2hex(random_bytes(16))
+                : $request->txid;
+
+            $statusId;
+
+            switch ($request->status) {
+                case 0:
+                    $statusId = Payment::STATUS_PENDING;
+                    break;
+                case 1:
+                    $statusId = Payment::STATUS_PENDING;
+                    break;
+                case 2:
+                    $statusId = Payment::STATUS_COMPLETED;
+                    break;
+                default:
+                    $statusId = Payment::STATUS_PENDING;
+            }
+
             $payment = $this->blockonomics->createPayment($data, $statusId);
+            $payment->custom_value1 = $request->txid;
+            $payment->custom_value2 = $request->btc_address;
+            $payment->custom_value3 = $request->btc_amount;
+            $payment->save();
```

Comment:
custom fields should not be used in the fashion. the transaction reference will capture the txid

you can use the private notes field to capture meta data

$payment->private_notes = "{$request->btc_address} - {$request->btc_amount}";

Reply:
Thanks for clarifying, now the changes have been implemented